pkg/networkservice/common/heal/client.go

```diff
@@ -70,61 +72,79 @@ func NewClient(client networkservice.MonitorConnectionClient, onHeal *networkser
 		updateExecutor:    serialize.NewExecutor(),
 		eventReceiver:     nil, // This is intentionally nil
 		recvEventExecutor: serialize.NewExecutor(),
+		chainContext:      ctx,
 	}
+
 	if rv.onHeal == nil {
 		rv.onHeal = addressof.NetworkServiceClient(rv)
 	}
-	rv.updateExecutor.AsyncExec(func() {
-		runtime.SetFinalizer(rv, func(f *healClient) {
-			f.updateExecutor.AsyncExec(func() {
-				if f.cancelFunc != nil {
-					f.cancelFunc()
-				}
-			})
-		})
-		ctx, cancelFunc := context.WithCancel(context.Background())
-		rv.cancelFunc = cancelFunc
-		// TODO decide what to do about err here
-		recv, _ := rv.client.MonitorConnections(ctx, nil)
-		rv.eventReceiver = recv
-		rv.recvEventExecutor.AsyncExec(rv.recvEvent)
-	})
+
+	rv.init()
 
 	return rv
 }
 
+func (f *healClient) init() {
+	if f.eventReceiver != nil {
+		select {
+		case <-f.eventReceiver.Context().Done():
+			return
+		default:
+			f.eventReceiver = nil
+		}
+	}
+
+	logrus.Info("Creating new eventReceiver")
+
+	recv, err := f.client.MonitorConnections(f.chainContext, &networkservice.MonitorScopeSelector{}, grpc.WaitForReady(true))
```

Comment:
Why are we Waiting For Ready here?

Reply:
When we receive `err` from `f.eventReceiver.Recv()` we assume that remote peer is silently died (without any "DELETE" events). In that case we call `init` again and method `MonitorConnection` should hangs for a while to let server UP and start listen.

Reply:
Ah... very slick, and since init is always run in an AsyncExec, the Request/Close calls don't block :)